### PR TITLE
[typing] add overload for empty `@task` decorator

### PR DESCRIPTION
**Pull request by zzstoatzz**
closes https://github.com/PrefectHQ/prefect/issues/16380

and more miscellaneous typing improvements to `prefect.tasks` as I was trying to identify the problem

### CODE REVIEW

File: src/prefect/tasks.py [hunk 1]
```diff
@@ -235,6 +226,14 @@ def __call__(self, parameters: dict[str, Any]) -> str:
         ...
 
 
+StateHookCallable: TypeAlias = Callable[
+    ["Task[..., Any]", TaskRun, State], Union[Awaitable[None], None]
+]
+TaskRunNameValueOrCallable: TypeAlias = Union[
+    Callable[[], str], TaskRunNameCallbackWithParameters, str
+]
```
Comment (zzstoatzz) [inline src/prefect/tasks.py:234]: this is unrelated to the linked issue, it was just common enough to back it out

File: src/prefect/tasks.py [hunk 2]
```diff
@@ -1774,34 +1769,31 @@ def task(
         if isinstance(__fn, (classmethod, staticmethod)):
             method_decorator = type(__fn).__name__
             raise TypeError(f"@{method_decorator} should be applied on top of @task")
-        return cast(
```
Comment (zzstoatzz) [inline src/prefect/tasks.py:1777]: we shouldn't need this `cast` anymore

File: tests/cli/test_deploy.py [hunk 3]
```diff
@@ -308,7 +308,7 @@ async def test_deploy_with_active_workers(
         await run_sync_in_worker_thread(
             invoke_and_assert,
             command=(
-                f"deploy ./wrapped-flow-project/flow.py:test_flow -n test-name -p {work_pool.name}"
+                f"deploy ./wrapped_flow_project/flow.py:test_flow -n test-name -p {work_pool.name}"
```
Comment (zzstoatzz) [inline tests/cli/test_deploy.py:311]: this directory name was giving `mypy` fits

File: src/prefect/tasks.py [hunk 4]
```diff
@@ -1585,27 +1580,33 @@ def task(__fn: Callable[P, R]) -> Task[P, R]:
     ...
 
 
+# see https://github.com/PrefectHQ/prefect/issues/16380
+@overload
+def task(
+    __fn: Literal[None] = None,
+) -> Callable[[Callable[P, R]], Task[P, R]]:
+    ...
```
Comment (zzstoatzz) [inline src/prefect/tasks.py:1588]: this is the fix

**Review by desertaxle — APPROVED** [review]
LGTM!
